File: Java_Joc_Windows_Android/AndroidStudioProject/core/src/base/menu/ComposedVisual.java
package base.menu;

import java.util.HashMap;
import java.util.Map;
import java.util.Vector;

import base.menu.util.ComputeSmartValuePoolData;
import local.app.AppLocal;
import base.visual.IDrawableNode;
import base.visual.IDynamicPropertiesObject;
import base.visual.PseudoVisualData;

/**
 * Created by Adrin on 6/4/2016.
 */
// IDrawableNode or Map<String, Object>;
public class ComposedVisual extends HashMap<String, Object> implements IComposedVisual {

    protected AppLocal app;

    // Only items created from templates are added to this list, manual items have to be manually added to this list.
    protected Vector<IComposedVisual> smartItemsList = new Vector<IComposedVisual>();

    private Map<String, Boolean> staticAnimatingPropertiesMap = new HashMap<String, Boolean>();

    private Map<String, Object> dataMap = new HashMap<String, Object>();

    protected Map<String, Object> createConfigMap = null;

    public ComposedVisual(AppLocal app) {
        super();

        this.app = app;
    }

    public void create(Map<String, Object> createConfigMap) {
        this.createConfigMap = createConfigMap;
    }

    public final void create() {
        create(null);
    }

    public Map<String, Object> getDataMap() {
        return dataMap;
    }

    public void setDataMapOverride(Map<String, Object> dataMap) {
        this.dataMap = dataMap;
    }

    public Object getItemNonNull(String[] itemID) {
        Object ret = getItem(itemID);

        if(ret == null) {
            app.errorHandler.handleAppCriticalError("WorldRoomLayers.1", "Error.");

            throw new Error("AnimatedComposedStiledVisual.Error.7.");
        }

        return ret;
    }
    public Object getItem(String[] itemID) {
        return getMenuItem(app,
                this, this,
                itemID, 0);
    }

    private static void recycleItem(AppLocal app, Object item, Vector<IComposedVisual> smartItemsList) {
        if (item instanceof IDrawableNode) {
            IDrawableNode cEntry = (IDrawableNode) item;

            if (cEntry.getParent() != null) {
                cEntry.getParent().removeChild(cEntry);
            }
        } else if(item instanceof IComposedVisual){
            IComposedVisual cEntry = (IComposedVisual) item;

            disposeItemsMap(app, cEntry, smartItemsList);
        }else if(item instanceof PseudoVisualData) {
            //void
        }else {
            app.log.postError("Error", "AnimatedComposedStiledVisual.Error.7.");

            throw new Error("AnimatedComposedStiledVisual.Error.7.");
        }
    }

    protected static synchronized Object getMenuItem(
            AppLocal app,
            IComposedVisual itemsMapRoot, IComposedVisual itemsMapLayer,
            String[] itemID, int currentItemIDIndex)
    {
        if(itemsMapRoot == null || itemsMapLayer == null) {
            app.errorHandler.handleAppCriticalError("WorldRoomLayers.1", "Error.");

            throw new Error("AnimatedComposedStiledVisual.Error.7.");
        }

        int _currentItemIDIndex = currentItemIDIndex;
        String currentIDLayer = itemID[_currentItemIDIndex];
        _currentItemIDIndex++;

        IComposedVisual _itemsMapLayer = null;
        if(currentIDLayer.equals("_local")) {
            _itemsMapLayer = itemsMapLayer;
        }else if(currentIDLayer.equals("_root")) {
            _itemsMapLayer = itemsMapRoot;
        }else if(_currentItemIDIndex == 1) {
            // The first item fxId entry must be a modifier.

            app.errorHandler.handleAppCriticalError("WorldRoomLayers.1", "Error.");

            throw new Error("AnimatedComposedStiledVisual.Error.7.");
        }else {
            Object entry = itemsMapLayer.getLocalItem(currentIDLayer);
            if (entry == null) {
                return null;
            }

            if(_currentItemIDIndex < itemID.length) {
                // This is not the last entry, a Map object is expected.
                if (!(entry instanceof IComposedVisual)) {
                    app.errorHandler.handleAppCriticalError("WorldRoomLayers.1", "Error.");

                    throw new Error("AnimatedComposedStiledVisual.Error.7.");
                }

                _itemsMapLayer = (IComposedVisual)entry;
            }else {
                return entry;
            }
        }

        if(itemID.length <= _currentItemIDIndex) {
            return _itemsMapLayer;
        }

        return getMenuItem(
                app,
                itemsMapRoot, _itemsMapLayer,
                itemID, _currentItemIDIndex);
    }

    public void setItem(String[] itemID, Object item) {
        setMenuItem(app,
                this, this,
                itemID, 0, item);
    }

    private static synchronized void setMenuItem(
            AppLocal app,
            IComposedVisual itemsMapRoot, IComposedVisual itemsMapLayer,
            String[] itemID, int currentItemIDIndex, Object item)
    {
        String[] itemLayerID = new String[itemID.length - 1];
        System.arraycopy(itemID, 0, itemLayerID, 0, itemLayerID.length);
        String itemIDFragment = itemID[itemID.length - 1];

        IComposedVisual itemLayer = (IComposedVisual)getMenuItem(app,
                itemsMapRoot, itemsMapLayer,
                itemLayerID, currentItemIDIndex);

        if(itemLayer == null) {
            app.errorHandler.handleAppCriticalError("WorldRoomLayers.1", "Error.");

            throw new Error("AnimatedComposedStiledVisual.Error.7.");
        }

        itemLayer.setLocalItem(itemIDFragment, item);
    }

    @Override
    public Object getLocalItem(String itemIDFragment) {
        return get(itemIDFragment);
    }

    @Override
    public void setLocalItem(String itemIDFragment, Object item) {
        Object existentItem = get(itemIDFragment);
        if(existentItem == null && item == null) {
            return ;
        }

        if(existentItem != null && item != null) {
            app.errorHandler.handleAppCriticalError("WorldRoomLayers.1", "Error.");

            throw new Error("AnimatedComposedStiledVisual.Error.7.");
        }

        if(existentItem != null) {
            put(itemIDFragment, null);

            if(existentItem instanceof IComposedVisual) {
                if(smartItemsList.indexOf(existentItem) < 0) {
                    app.errorHandler.handleAppCriticalError("WorldRoomLayers.1", "Error.");

                    throw new Error("AnimatedComposedStiledVisual.Error.7.");
                }

                smartItemsList.remove((IComposedVisual)existentItem);
            }
        }

        if(item != null) {
            put(itemIDFragment, item);

            if(!itemIDFragment.equals("parent") && item instanceof IComposedVisual) {
                if(smartItemsList.indexOf(item) >= 0) {
                    app.errorHandler.handleAppCriticalError("WorldRoomLayers.1", "Error.");

                    throw new Error("AnimatedComposedStiledVisual.Error.7.");
                }

                smartItemsList.add((IComposedVisual)item);
            }
        }
    }

    public Object getItemPropertyValue(String[] itemID, String property, ComputeSmartValuePoolData poolData) {
        return getVisualPropertyValue(app, this, this, itemID, property, null, poolData, dataMap);
    }

    protected static Object getVisualPropertyValue(AppLocal app, IComposedVisual itemsMap, IComposedVisual itemsMapLayer, String[] itemID, String property, IDynamicPropertiesObject selfObject, ComputeSmartValuePoolData poolData, Map<String, Object> dataMap) {
        IDynamicPropertiesObject visualItem = null;

        String firstIDLayer = itemID[0];
        if(firstIDLayer.equals("_appSize")) {
            if(poolData != null && (poolData.isPropDependsOnAnimating == null || poolData.isPropDependsOnAnimating != true)) {
                poolData.isPropDependsOnAnimating = false;
            }

            if(property.equals("width")) {
                return app.screenSize.width;
            }else if(property.equals("height")) {
                return app.screenSize.height;
            }else {
                app.errorHandler.handleAppCriticalError("WorldRoomLayers.1", "Error.");

                throw new Error("AnimatedComposedStiledVisual.Error.7.");
            }
        }else if(firstIDLayer.equals("_dpm")) {
            if(poolData != null && (poolData.isPropDependsOnAnimating == null || poolData.isPropDependsOnAnimating != true)) {
                poolData.isPropDependsOnAnimating = false;
            }

            if(property.equals("width")) {
                return app.screenDensityMultiplier.x;
            }else if(property.equals("height")) {
                return app.screenDensityMultiplier.y;
            }else {
                app.errorHandler.handleAppCriticalError("WorldRoomLayers.1", "Error.");

                throw new Error("AnimatedComposedStiledVisual.Error.7.");
            }
        }else if(firstIDLayer.equals("_self")) {
            if(selfObject == null) {
                app.errorHandler.handleAppCriticalError("WorldRoomLayers.1", "Error.");

                throw new Error("AnimatedComposedStiledVisual.Error.7.");
            }

            visualItem = selfObject;
        }else if(firstIDLayer.equals("_dataMap")) {
            if(poolData != null && (poolData.isPropDependsOnAnimating == null || poolData.isPropDependsOnAnimating != true)) {
                poolData.isPropDependsOnAnimating = true;
            }

            if(dataMap == null) {
                app.errorHandler.handleAppCriticalError("WorldRoomLayers.1", "Error.");

                throw new Error("AnimatedComposedStiledVisual.Error.7.");
            }

            return dataMap.get(property);
        }else {
            try {
                visualItem = (IDynamicPropertiesObject)getMenuItem(app, itemsMap, itemsMapLayer, itemID, 0);
            }catch(ClassCastException x) {
                app.errorHandler.handleAppCriticalError("WorldRoomLayers.1", "Error.");

                throw new Error("AnimatedComposedStiledVisual.Error.7.");
            }
        }

        if(visualItem == null) {
            app.errorHandler.handleAppCriticalError("WorldRoomLayers.1", "Error.");

            throw new Error("AnimatedComposedStiledVisual.Error.7.");
        }

        float ret;

        if(poolData != null && (poolData.isPropDependsOnAnimating == null || poolData.isPropDependsOnAnimating != true)) {
            Boolean b = visualItem.getStaticAnimatingPropertyStatus(property);
            if (b == null) {
                app.errorHandler.handleAppCriticalError("WorldRoomLayers.1", "Error.");

                throw new Error("AnimatedComposedStiledVisual.Error.7.");
            }

            poolData.isPropDependsOnAnimating = b;
        }
        try {
            ret = (Float) visualItem.getPropertyValue(property);
        }catch(ClassCastException x) {
            app.errorHandler.handleAppCriticalError("WorldRoomLayers.1", "Error.");

            throw new Error("AnimatedComposedStiledVisual.Error.7.");
        }catch(NullPointerException x) {
            app.errorHandler.handleAppCriticalError("WorldRoomLayers.1", "Error.");

            throw new Error("AnimatedComposedStiledVisual.Error.7.");
        }

        return ret;
    }

    private static void disposeItemsMap(AppLocal app, IComposedVisual entriesMap, Vector<IComposedVisual> smartItemsList) {
        for (Map.Entry<String, Object> entry : entriesMap.entrySet()) {
            if(entry != null && entry.getValue() != null && !entry.getKey().equals("parent")) {
                Object item = entry.getValue();

                if(item instanceof IComposedVisual) {
                    ((IComposedVisual)item).dispose();

                    smartItemsList.remove((IComposedVisual)item);
                }

                recycleItem(app, item, smartItemsList);
            }
        }
    }

    public void appLoop(float deltaS, int deltaMS) {
        // Call app loop on items that use it and implement IAppLoopItem.
        for(int i=0;i<smartItemsList.size();i++) {
            IComposedVisual entry = smartItemsList.get(i);
            if(entry != null) {
                entry.appLoop(deltaS, deltaMS);
            }
        }
    }

    public Boolean getStaticAnimatingPropertyStatus(String property) {
        return staticAnimatingPropertiesMap.get(property);
    }

    public void putStaticAnimatingPropertyStatus(String property, Boolean status) {
        staticAnimatingPropertiesMap.put(property, status);
    }

    public void dispose() {
        disposeItemsMap(app, this, smartItemsList);
    }

}

File: CPP_Joc_Windows_Android/SH_Client_Win_Cpp_Cmake/App/src/base/menu/ComposedVisual.cpp
#include "ComposedVisual.h"
#include <base/exceptions/LogicException.h>
#include <graphics/visual2d/drawable/IDrawableNode2D.h>
#include <graphics/visual/drawable/PseudoVisualData.h>
#include <graphics/visual2d/drawable/IContainer2D.h>
#include <base/app/IAppMetrics.h>
#include <base/component/IComponent_UIView.h>

ComposedVisual::ComposedVisual(IAppMetrics* appMetrics)
    : super(IComponent_UIView::COMPONENT_CATEGORY),
    appMetrics(appMetrics)
{
    //void
}

void ComposedVisual::setCreateConfigMap(Map1D<std::string, void*>* createConfigMap) {
    this->createConfigMap = createConfigMap;
}

void ComposedVisual::create() {
    super::create();

    createFramework();
    createContent();
}

void ComposedVisual::createFramework() {
    //void
}

void ComposedVisual::createContent() {
    //void
}

IMap1D<std::string, std::shared_ptr<base::IWrappedValue>>* ComposedVisual::getDataMap() {
    return &dataMap;
}

void ComposedVisual::putDataMapValue(std::string& key, std::shared_ptr<base::IWrappedValue> val, bool doAnimatingVisualsSync) {
    dataMap.putReference(key, val);
}

void ComposedVisual::putDataMapValueB(std::string key, std::shared_ptr<base::IWrappedValue> val, bool doAnimatingVisualsSync) {
	putDataMapValue(key, val, doAnimatingVisualsSync);
}

void ComposedVisual::addDataMap(Map1D<std::string, std::shared_ptr<base::IWrappedValue>>* dataMap) {
    this->dataMap.addMap(dataMap);
}

void ComposedVisual::recycleItem(IGenericDrawable* item, ArrayList<IComposedVisual*>* smartItemsList) {
    IDrawableNode2D* cEntry_dn2d;
    IComposedVisual* cEntry_cv;
    PseudoVisualData* cEntry_pvd;
    if ((cEntry_dn2d = dynamic_cast<IDrawableNode2D*>(item))) {
        if (cEntry_dn2d->getParent() != nullptr) {
            cEntry_dn2d->getParent()->removeChild(cEntry_dn2d);
        }
    } else if((cEntry_cv = dynamic_cast<IComposedVisual*>(item))){
        //void
    }else if((cEntry_pvd = dynamic_cast<PseudoVisualData*>(item))) {
        //void
    }else {
        throw LogicException(LOC);
    }
}

IGenericDrawable* ComposedVisual::getMenuItemOptional(
        IComposedVisual* itemsMapRoot, IComposedVisual* itemsMapLayer,
        UiId& composedId, int currentItemIDIndex)
{
    if(itemsMapRoot == nullptr || itemsMapLayer == nullptr) {
        throw LogicException(LOC);
    }

    int _currentItemIDIndex = currentItemIDIndex;
    std::string& currentIDLayer = composedId.getSplitId()->getReference(_currentItemIDIndex);
    _currentItemIDIndex++;

    IComposedVisual* _itemsMapLayer = nullptr;
    if(currentIDLayer == "_local") {
        _itemsMapLayer = itemsMapLayer;
    }else if(currentIDLayer == "_root") {
        _itemsMapLayer = itemsMapRoot;
    }else if(_currentItemIDIndex == 1) {
        // The first item fxId listEntry must be a modifier.

        throw LogicException(LOC);
    }else {
        IGenericDrawable* entry = itemsMapLayer->getLocalItem(currentIDLayer);
        if (entry == nullptr) {
            return nullptr;
        }

        if(_currentItemIDIndex < composedId.getSplitId()->size()) {
            // This is not the last listEntry, a Map object is expected.
            _itemsMapLayer = dynamic_cast<IComposedVisual*>(entry);
            if (!_itemsMapLayer) {
                throw LogicException(LOC);
            }
        }else {
            return entry;
        }
    }

    if(composedId.getSplitId()->size() <= _currentItemIDIndex) {
        return _itemsMapLayer;
    }

    return getMenuItemOptional(
            itemsMapRoot, _itemsMapLayer,
            composedId, _currentItemIDIndex);
}

std::shared_ptr<base::IWrappedValue> ComposedVisual::getVisualPropertyValue(IComposedVisual* itemsMap, IComposedVisual* itemsMapLayer, UiId& itemComposedID, std::string& property, IDynamicPropertiesObject2D* selfObject, ComputeSmartValuePoolData* poolData, IMap1D<std::string, std::shared_ptr<base::IWrappedValue>>* dataMap, IAppMetrics* appMetrics) {
    IDynamicPropertiesObject2D* visualItem = nullptr;

    std::string firstIDLayer = itemComposedID.getSplitId()->getDirect(0);
    if(firstIDLayer == "_appSize") {
        if(poolData != nullptr && (!poolData->isInit_isPropDependingOnAnimating() || poolData->isTrue_isPropDependingOnAnimating() != true)) {
            poolData->set_isPropDependsOnAnimating(false);
        }

        if(property == "width") {
            return base::IWrappedValue::new_float(appMetrics->getScreen()->getWindowResolutionX());
        }else if(property == "height") {
            return base::IWrappedValue::new_float(appMetrics->getScreen()->getWindowResolutionY());
        }else {
            throw LogicException(LOC);
        }
    }else if(firstIDLayer == "_dpm") {
        if(poolData != nullptr && (!poolData->isInit_isPropDependingOnAnimating() || poolData->isTrue_isPropDependingOnAnimating() != true)) {
            poolData->set_isPropDependsOnAnimating(false);
        }

        if(property == "width") {
            return base::IWrappedValue::new_float(appMetrics->getScreen()->getScreenDensityMultiplier()->x);
        }else if(property == "height") {
            return base::IWrappedValue::new_float(appMetrics->getScreen()->getScreenDensityMultiplier()->y);
        }else {
            throw LogicException(LOC);
        }
    }else if(firstIDLayer == "_dpm_font") {
        if (poolData != nullptr && (!poolData->isInit_isPropDependingOnAnimating() || poolData->isTrue_isPropDependingOnAnimating() != true)) {
            poolData->set_isPropDependsOnAnimating(false);
        }
        
        return base::IWrappedValue::new_float(appMetrics->getScreen()->getScreenDensityMultiplier_fonts());
    }else if(firstIDLayer == "_tSize") {
        if(poolData != nullptr && (!poolData->isInit_isPropDependingOnAnimating() || poolData->isTrue_isPropDependingOnAnimating() != true)) {
            poolData->set_isPropDependsOnAnimating(false);
        }

        if(property == "width") {
            return base::IWrappedValue::new_float(appMetrics->getScreen()->getWindowResolutionX() / appMetrics->getScreen()->getTemplateSize()->x);
        }else if(property == "height") {
            return base::IWrappedValue::new_float(appMetrics->getScreen()->getWindowResolutionY() / appMetrics->getScreen()->getTemplateSize()->y);
        }else {
            throw LogicException(LOC);
        }
    }else if(firstIDLayer == "_self") {
        if(selfObject == nullptr) {
            throw LogicException(LOC);
        }

        visualItem = selfObject;
    }else if(firstIDLayer == "_dataMap") {
        if(poolData != nullptr && (!poolData->isInit_isPropDependingOnAnimating() || poolData->isTrue_isPropDependingOnAnimating() != true)) {
            poolData->set_isPropDependsOnAnimating(true);
        }

        if(dataMap == nullptr) {
            throw LogicException(LOC);
        }

        return dataMap->getDirect(property);
    }else if(firstIDLayer == "_standardIncrement") {
        if(poolData != nullptr && (!poolData->isInit_isPropDependingOnAnimating() || poolData->isTrue_isPropDependingOnAnimating() != true)) {
            poolData->set_isPropDependsOnAnimating(false);
        }

        return base::IWrappedValue::new_float(appMetrics->getScreen()->getStandardIncrement());
    }else {
        visualItem = dynamic_cast<IDynamicPropertiesObject2D*>(getMenuItemOptional(itemsMap, itemsMapLayer, itemComposedID, 0));
        if(visualItem == nullptr) {
            // Check these values when a error occurs. This ussually occurs when attempting to access a object that doesn't exist.
            std::string s0 = itemComposedID.getSplitId()->getReference(0);
            std::string s1 = itemComposedID.getSplitId()->getReference(1);

            throw LogicException(LOC);
        }
    }

    if(visualItem == nullptr) {
        throw LogicException(LOC);
    }

    std::shared_ptr<base::IWrappedValue> ret;

    if(poolData != nullptr && (!poolData->isInit_isPropDependingOnAnimating() || poolData->isTrue_isPropDependingOnAnimating() != true)) {
        if (!visualItem->isInit_StaticAnimatingPropertyStatus(property)) {
            throw LogicException(LOC);
        }

        bool b = visualItem->isTrue_StaticAnimatingPropertyStatus(property);
        poolData->set_isPropDependsOnAnimating(b);
    }
    ret = visualItem->getPropertyValue(property);

    return ret;
}

void ComposedVisual::setItem(UiId& composedId, IGenericDrawable* item) {
    setMenuItem(
        this, this,
        composedId, 0, item);
}

void ComposedVisual::setItem(std::string& composedId, IGenericDrawable* item) {
    UiId* id = UiId::newIdFromComposedId(composedId, "_root");

    setItem(*id, item);

    delete id;
}

void ComposedVisual::setItemB(std::string composedId, IGenericDrawable* item) {
	setItem(composedId, item);
}

void ComposedVisual::setMenuItem(
        IComposedVisual* itemsMapRoot, IComposedVisual* itemsMapLayer,
        UiId& composedId, int currentItemIDIndex, IGenericDrawable* item)
{
    UiId subComposedId(composedId, 0, composedId.getSplitId()->size() - 1);

    std::string& itemID = composedId.getSplitId()->getReference(composedId.getSplitId()->size() - 1);

    IComposedVisual* itemLayer = dynamic_cast<IComposedVisual*>(getMenuItemOptional(
            itemsMapRoot, itemsMapLayer,
            subComposedId, currentItemIDIndex));

    if(itemLayer == nullptr) {
        throw LogicException(LOC);
    }

    itemLayer->setLocalItem(itemID, item);
}

IGenericDrawable* ComposedVisual::getLocalItem(std::string itemIDFragment) {
    return subItemsMap.getDirect(itemIDFragment, nullptr);
}

void ComposedVisual::setLocalItem(std::string itemIDFragment, IGenericDrawable* item) {
    IGenericDrawable* existentItem = subItemsMap.getDirect(itemIDFragment, nullptr);
    if(existentItem == nullptr && item == nullptr) {
        return ;
    }

    if(existentItem != nullptr && item != nullptr) {
        throw LogicException(LOC);
    }

    if(existentItem != nullptr) {
        subItemsMap.putDirect(itemIDFragment, nullptr);

        IComposedVisual* existentItem_icv;
        if((existentItem_icv = dynamic_cast<IComposedVisual*>(existentItem))) {
            if(smartItemsList.indexOf(existentItem_icv) < 0) {
                throw LogicException(LOC);
            }

            smartItemsList.remove(existentItem_icv);
        }
    }

    if(item != nullptr) {
        subItemsMap.putDirect(itemIDFragment, item);

        IComposedVisual* item_icv;
        if(itemIDFragment != "parent" && (item_icv = dynamic_cast<IComposedVisual*>(item))) {
            if(smartItemsList.indexOf(item_icv) >= 0) {
                throw LogicException(LOC);
            }

            smartItemsList.appendDirect(dynamic_cast<IComposedVisual *>(item));
        }
    }
}

void ComposedVisual::disposeItemsMap(IComposedVisual* entriesMap, ArrayList<IComposedVisual*>* smartItemsList) {
    for (std::unordered_map<std::string, IGenericDrawable*>::iterator it = entriesMap->getSubItemsMap()->map.begin(); it != entriesMap->getSubItemsMap()->map.end(); ++it) {
        if(it->second != nullptr && it->first != "parent") {
            IGenericDrawable* item = it->second;

            IComposedVisual* item_icv;
            if((item_icv = dynamic_cast<IComposedVisual*>(item))) {
                item_icv->reservedDisposeMain();

                smartItemsList->remove(item_icv);
            }

            recycleItem(item, smartItemsList);
        }
    }
}

void ComposedVisual::visualLoop(float deltaS, int deltaMS) {
    // Call app loop on items that use it and implement IAppLoopItem.
	// Call children first even tough this can set their properties because this may use values that are already updated in the children.
	// visualLoopB will ensure the children won't have any update delay.
    for(int i=0;i<smartItemsList.size();i++) {
        IComposedVisual* entry = smartItemsList.getDirect(i);
        if(entry != nullptr) {
            entry->visualLoop(deltaS, deltaMS);
        }
    }

    for (ListDL<IComposedVisual*>::Entry* entry = manualSmartItemsList.getFirst(); entry != nullptr; entry = entry->next) {
        IComposedVisual* entryData = entry->data;

        entryData->visualLoop(deltaS, deltaMS);
    }
}

void ComposedVisual::visualLoopB(float deltaS, int deltaMS) {
    // Call app loop on items that use it and implement IAppLoopItem.
	// Call children first even tough this can set their properties because this may use values that are already updated in the children.
	// The children may be updated 2 times in certain cases (eg. window size change) and have a 1 frame update delay but that is ok.
    for(int i=0;i<smartItemsList.size();i++) {
        IComposedVisual* entry = smartItemsList.getDirect(i);
        if(entry != nullptr) {
            entry->visualLoopB(deltaS, deltaMS);
        }
    }

    for (ListDL<IComposedVisual*>::Entry* entry = manualSmartItemsList.getFirst(); entry != nullptr; entry = entry->next) {
        IComposedVisual* entryData = entry->data;

        entryData->visualLoopB(deltaS, deltaMS);
    }
}

bool ComposedVisual::isInit_StaticAnimatingPropertyStatus(std::string& property) {
	if (staticAnimatingPropertiesMap.getDirect(property, -1) != -1) {
		return true;
	}

	return false;
}

bool ComposedVisual::isTrue_StaticAnimatingPropertyStatus(std::string& property) {
	int v = staticAnimatingPropertiesMap.getDirect(property, -1);

	if (v == 1) {
		return true;
	} else if (v == 0) {
		return false;
	}

	throw LogicException(LOC);
}

bool ComposedVisual::isTrue_StaticAnimatingPropertyStatusDirect(std::string property) {
	return isTrue_StaticAnimatingPropertyStatus(property);
}

void ComposedVisual::putStaticAnimatingPropertyStatus(std::string& property, int status) {
	staticAnimatingPropertiesMap.putDirect(property, status);
}

void ComposedVisual::putStaticAnimatingPropertyStatusDirect(std::string property, int status) {
	putStaticAnimatingPropertyStatus(property, status);
}

std::shared_ptr<base::IWrappedValue> ComposedVisual::getItemPropertyValue(UiId& composedId, std::string& property, ComputeSmartValuePoolData* poolData) {
    return getVisualPropertyValue(this, this, composedId, property, nullptr, poolData, &dataMap, appMetrics);
}

std::shared_ptr<base::IWrappedValue> ComposedVisual::getItemPropertyValue(std::string smartItemIdAndProperty, ComputeSmartValuePoolData* poolData) {
    UiId cId{smartItemIdAndProperty, "_root", true};
    std::string prop = UiId::extractPropStringFromComposedIdProp(smartItemIdAndProperty);

    return getItemPropertyValue(
        cId,
        prop,
        poolData
    );
}

void ComposedVisual::disposeMain() {
    disposeItemsMap(this, &smartItemsList);

    super::disposeMain();
}

ComposedVisual::~ComposedVisual() {
    //void
}
